### PR TITLE
Address feedback for `Fabric.init_module()` (4/4)

### LINKED ISSUE #17581: RFC: Future of init context managers in Fabric
### Description & Motivation

Recently, #17488 was merged with the following changes in summary:

- `Fabric.init()`: managing the device and precision part that Fabric.init_module() does
- Undeprecated `fabric.sharded_model()`
- Renamed the `Strategy.init_module` and `Precision.init_module` to just `init_context()`
- Renamed `Strategy.init_sharded_context` without backward compatibility

These points work against the preparation work done in #17462 to provide a unified, strategy-agnostic way of initializing the model, and the PR didn't provide much motivation to for this overhaul. The main point was given in the documentation: "Using these separately can provide more control for expert users".

Here are my concerns and why #17488 should be reverted, or partially reverted.

**Concern 1: There are multiple ways of doing the same thing now.**
`Fabric.init()`, `Fabric.init_module()` have overlapping functionality. They can both be used to initialize the module and control how tensors are created. "Which one should be used?" is what the user could ask. The ambiguous name "init" doesn't help here. Furthermore, `Fabric.init()`'s implementation is tied to the strategy because it calls `Strategy.init_context()`, but why should the strategy decide how tensors are created outside the model? There is a lack of motivation through examples and use cases. From a documentation perspective, personally, I would find myself struggling to explain the differences to a newcomer.

**Concern 2: `Fabric.init()` is redundant.**
PyTorch accepts device and dtype attributes everywhere and has plenty of flexibility when it comes to tensor creation this way. With the `fabric.device` property, you can already create tensors directly on the device. In my opinion, this is an unneeded PyTorch abstraction layer. The cost is that it increases the API surface of Fabric, and makes the user study the differences between the overlapping features of the context managers. It is unclear to me why this feature was promoted as "more flexibility for expert users" when in reality the choice of dtype and device is tied to the strategy. 

**Concern 3: `Fabric.init` is a misleading name**
Generally speaking, method names should be communicating the action they perform on the object or what they are meant to return. Here, without knowing any further context, `Fabric.init()` communicates that it initializes Fabric itself, which is of course wrong. The previous name `Fabric.init_module()` was better, because the user is forced to study its return type to understand the intentions. You can't pass in the model as an argument, because that would lead to an error and thus the user will understand that this is actually a context manager.

**Concern 4: The `sharded_model` API is misleading and misunderstood.**
One motivation behind the original PR #17462 was to provide a better-named API for instantiating modules, where both the name and the functionality would be agnostic to the strategy/device choice. This meant the `Fabric.sharded_model()` context had to be discouraged in favor of the more generalized `Fabric.init_module()`. Confusions like https://github.com/Lightning-AI/lightning/pull/17488#pullrequestreview-1414249897 are the prime example that there exists a misunderstanding of what `Fabric.sharded_model()` does. Hint: It does NOT shard the model!

**Concern 5: It is unclear where empty weight initialization /  deferred init should be implemented now.**
The motivation in #17462 was to have a unified model instantiation context that would take care of efficient initialization, and potentially deferred/lazy instantiation if it made sense given the strategy. Where should this feature be added now? To only some of the context managers, to all of them? It is very unclear to me what the future of this feature is now. 

### Pitch

My proposal to address the concerns is:

1. Discuss the above concerns and align on the future to provide a more focused API.
2. Discourage the use of the `sharded_model()` by reverting that part in #17488. 
3. Either revert the addition of `Fabric.init()` or keep it at least undocumented/protected until there is a strong demand for it.

### Alternatives

A) An alternative would be to find better-fitting names for the context managers discussed here, but it would only address the concerns partially. There would still be a lot of open questions. 

B) Instead of `Fabric.init()`, provide a standalone function/context manager (pseudo code):

```py
@contextmanager
def init(device, dtype):
    torch.set_default_tensor_dtype(dtype)
    with device:
        yield
    ...
```

### Additional context

Requesting review from @carmocca @lantiga @justusschock 

We should align on this before 2.1, because after that it will be too late to change any of these APIs.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #17581 
Partially reverts #17488

- Keeps `Fabric.init()` disentangled from `Fabric.init_module()`
- Renames `Fabric.init()` to `Fabric.init_tensor()`

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -584,9 +584,12 @@ def no_backward_sync(self, module: _FabricModule, enabled: bool = True) -> Gener
...
     @contextmanager
     def sharded_model(self) -> Generator:
-        """Shard the parameters of the model instantly when instantiating the layers."""
-        context = self._strategy.init_sharded_context() if isinstance(self._strategy, _Sharded) else nullcontext()
-        with context:
+        """Instantiate a model under this context manager to prepare it for model-parallel sharding.
+
+        .. deprecated:: This context manager is deprecated in favor of :meth:`init_module`, use it instead.
+        """
+        rank_zero_deprecation("`Fabric.sharded_model()` is deprecated in favor of `Fabric.init_module()`.")
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:591]: Since we agree on adding back this deprecation, it could be pulled out into a separate PR to get it merged sooner

File: src/lightning/fabric/fabric.py [hunk 2]
```diff
@@ -596,16 +599,27 @@ def init(self) -> Generator:
         The parameters get created on the device (if using PyTorch 2.0 or newer) and with the right data type right away
         without wasting memory being allocated unnecessarily.
         """
-        with self._strategy.init_context():
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/fabric/fabric.py:599]: Because the strategy in `module_init_context()` does more than just set the dtype and device. For example, we don't want that the deepspeed zero Init() context manager is being used when we do

```py
with fabric.init():
    torch.ones(2, 2)
```

or other side effects from the strategy. The two IMO features should be completely disentangled from each other.

File: src/lightning/fabric/fabric.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/fabric.py:591]: Please review this: https://github.com/Lightning-AI/lightning/pull/17721

**Comment by awaelchli** [comment]
This is blocking my work on the feature. Please review #17721 and the other parts asap.

File: src/lightning/fabric/fabric.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/fabric.py:591]: Done!

File: src/lightning/fabric/fabric.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/fabric/fabric.py:609]: Let me phrase it in another way. I want that
1. init_tensor() is for tensors only (device + dtype). For this, we don't need an extra hook on the strategy because it's just a simple with `self.device` call that generalizes for all tensors.
2. init_module() is for module instantiation only. should do whatever the strategy wants. This deserves its own hook where the strategy can put their own context managing logic. For the non-distributed strategies, the code is trivial and there is no difference to what init_tensor() does. But for deepspeed and fsdp, we need to have custom logic. 

I want to disentangle the two completely (I insist!) so that I can implement empty weight initialization in #17616 in `init_module()` only.

**Review by justusschock — APPROVED** [review]

File: src/lightning/fabric/strategies/deepspeed.py [hunk 3]
```diff
@@ -339,6 +339,11 @@ def setup_optimizer(self, optimizer: Optimizer) -> Optimizer:
         """
         raise NotImplementedError(self._err_msg_joint_setup_required())
 
+    @contextmanager
+    def module_init_context(self) -> Generator[None, None, None]:
+        with self.module_sharded_context():
```
Comment (carmocca) [inline src/lightning/fabric/strategies/deepspeed.py:344]: Isn't this missing the precision call?

**Review by carmocca — APPROVED** [review]
Approving. Discussing further offline

File: src/lightning/fabric/strategies/deepspeed.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/deepspeed.py]: Yes to some degree. DeepSpeed's Zero context manager handles precision already. So in theory we should put the precision context inside the else that mirrors `self.zero_stage_3` in L355. For now, I added it conditionally in the init_module method, but in the future, we may want to rearrange the logic for better readability.
